sql/wsrep_schema.cc

```diff
@@ -539,38 +540,14 @@ Wsrep_schema::Wsrep_schema() {}
 
 Wsrep_schema::~Wsrep_schema() {}
 
-static void wsrep_init_thd_for_schema(THD *thd) {
-  thd->security_context()->skip_grants();
-  thd->system_thread = SYSTEM_THREAD_BACKGROUND;
-
-  thd->real_id = pthread_self();  // Keep purify happy
-
-  thd->set_time();
-
-  /* */
-  thd->variables.wsrep_on = 0;
-  /* No binlogging */
-  thd->variables.sql_log_bin = 0;
-  thd->variables.option_bits &= ~OPTION_BIN_LOG;
-  /* No general log */
-  thd->variables.option_bits |= OPTION_LOG_OFF;
-  /* Read committed isolation to avoid gap locking */
-  thd->variables.transaction_isolation = ISO_READ_COMMITTED;
-  wsrep_assign_from_threadvars(thd);
-  wsrep_store_threadvars(thd);
-}
-
-int Wsrep_schema::init() {
+bool Wsrep_schema::init(THD *thd) {
   DBUG_ENTER("Wsrep_schema::init()");
-  int ret;
-  THD *thd = new THD;
-  thd->set_new_thread_id();
-  if (!thd) {
-    WSREP_ERROR("Unable to get thd");
-    DBUG_RETURN(1);
+
+  handlerton *ddse = ha_resolve_by_legacy_type(thd, DB_TYPE_INNODB);
```

Comment:
Do we need it here again? It seems that caller does the same.

Reply:
Yes, if I remember correctly it called from somewhere else too - and without adding it to the called, we would create the object, but wouldn't initialize it, which is again an issue.

Reply:
Declaration of wsrep_schema::init is in wsrep_schema.h. It changed - thd parameter was added. The only place I see its invocation changed is in caller - wsrep_init_schema. So it can not be called from any other place.

Reply:
Fixed, it's really enough at one place. The failure I was seeing was related to the missing DBUG_RETURN noticed by Kenn.